### PR TITLE
Added Assessment step to estimate the size of DBFS Root Tables.

### LINKED ISSUE #330: Compute size of `MANAGED` tables on DBFS root
Add a task to assessment workflow right after `crawl_tables` to compute table size:

# + show tblproperties db.table
spark._jsparkSession.table("db.table").queryExecution().analyzed().stats().sizeInBytes()

### PR DESCRIPTION
**Pull request by FastLee**
Closes #330 

- Added assessment step estimate dbfs root table size.
- This should help deciding whether the table is a good candidate for cloning.

### CODE REVIEW

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 1]
```diff
@@ -198,12 +216,12 @@ def _describe(self, catalog: str, database: str, table: str) -> Table | None:
...
 class TablesMigrate:
     def __init__(
-        self,
-        tc: TablesCrawler,
-        ws: WorkspaceClient,
-        backend: SqlBackend,
-        default_catalog=None,
-        database_to_catalog_mapping: dict[str, str] | None = None,
+            self,
+            tc: TablesCrawler,
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/tables.py:220]: make fmt

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/hive_metastore/table_size.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+@dataclass
+class TableSize:
+    catalog: str
+    database: str
+    name: str
+    size_in_bytes: int
+
+
+class TableSizeCrawler(CrawlerBase):
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_size.py:20]: tables.py is getting too big. We split all the others to separate files

File: src/databricks/labs/ucx/hive_metastore/table_size.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    def __init__(self, backend: SqlBackend, schema):
+        """
+        Initializes a TablesSizeCrawler instance.
+
+        Args:
+            backend (SqlBackend): The SQL Execution Backend abstraction (either REST API or Spark)
+            schema: The schema name for the inventory persistence.
+        """
+        super().__init__(backend, "hive_metastore", schema, "table_size", TableSize)
+        self._table_crawler = TablesCrawler(backend, schema)
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_size.py:31]: Fixed

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 4]
```diff
@@ -80,6 +80,23 @@ def sql_unset_upgraded_to(self, catalog):
...
+    def is_dbfs_root(self) -> bool:
+        if not self.location:
+            return False
+        if (
+                self.location.startswith("/dbfs/mnt")
+                or
+                self.location.startswith("dbfs:/mnt")
+        ):
+            return False
+        if (
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/tables.py]: I don't understand. I split it to three if statements.

File: tests/unit/framework/mocks.py [hunk 5]
```diff
@@ -52,3 +53,14 @@ def rows_written_for(self, full_name: str, mode: str) -> list[DataclassInstance]
                 continue
             rows += stub_rows
         return rows
+
+
+class MockRuntimeBackend(MockBackend):
```
Comment (nfx) [inline tests/unit/framework/mocks.py:58]: no need to do this. remove.

see https://github.com/databrickslabs/ucx/blob/main/tests/unit/framework/test_crawlers.py#L167-L246

File: src/databricks/labs/ucx/framework/crawlers.py [hunk 6]
```diff
@@ -167,6 +168,10 @@ def save_table(self, full_name: str, rows: Sequence[DataclassInstance], klass: D
         df = self._spark.createDataFrame(rows, self._schema_for(klass))
         df.write.saveAsTable(full_name, mode=mode)
 
+    def get_table_size(self, table_full_name: str) -> int:
+        logger.debug(f"Evaluating {table_full_name} table size.")
+        return self._spark._jsparkSession.table(f"{table_full_name}").queryExecution().analyzed().stats().sizeInBytes()
```
Comment (nfx) [inline src/databricks/labs/ucx/framework/crawlers.py:173]:
```suggestion
        return self._spark._jsparkSession.table(table_full_name).queryExecution().analyzed().stats().sizeInBytes()
```

File: src/databricks/labs/ucx/framework/crawlers.py [hunk 7]
```diff
@@ -147,7 +148,7 @@ def __init__(self):
             msg = "Not in the Databricks Runtime"
             raise RuntimeError(msg)
 
-        self._spark = SparkSession.builder.getOrCreate()
+        self._spark = create_autospec(SparkSession)
```
Comment (nfx) [inline src/databricks/labs/ucx/framework/crawlers.py:151]: revert

File: src/databricks/labs/ucx/framework/crawlers.py [hunk 8]
```diff
@@ -167,6 +168,10 @@ def save_table(self, full_name: str, rows: Sequence[DataclassInstance], klass: D
         df = self._spark.createDataFrame(rows, self._schema_for(klass))
         df.write.saveAsTable(full_name, mode=mode)
 
+    def get_table_size(self, table_full_name: str) -> int:
```
Comment (nfx) [inline src/databricks/labs/ucx/framework/crawlers.py:171]: doesn't belong here. move to `TableSizeCrawler`.

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 9]
```diff
@@ -80,6 +80,15 @@ def sql_unset_upgraded_to(self, catalog):
             f"UNSET TBLPROPERTIES IF EXISTS('upgraded_to');"
         )
 
+    def is_dbfs_root(self) -> bool:
+        if not self.location:
+            return False
+        if self.location.startswith("/dbfs/mnt") or self.location.startswith("dbfs:/mnt"):
+            return False
+        if self.location.startswith("/dbfs/") or self.location.startswith("dbfs:/"):
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/tables.py:88]: what about databricks datasets?

File: src/databricks/labs/ucx/framework/crawlers.py [hunk 10]
```diff
@@ -6,6 +6,7 @@
 from collections.abc import Callable, Iterable, Iterator, Sequence
 from types import UnionType
 from typing import Any, ClassVar, Generic, Protocol, TypeVar
+from unittest.mock import create_autospec
```
Comment (nfx) [inline src/databricks/labs/ucx/framework/crawlers.py:9]: revert

File: src/databricks/labs/ucx/hive_metastore/table_size.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+    def _crawl(self) -> Iterable[TableSize]:
+        """Crawls and lists tables using table crawler
+        Identifies DBFS root tables and calculates the size for these.
+        """
+        table_sizes = []
+        for table in self._table_crawler.snapshot():
+            if not table.kind == "TABLE" or not table.is_dbfs_root():
+                continue
+            table_size = TableSize(catalog=table.catalog, database=table.database, name=table.name, size_in_bytes=0)
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_size.py]: Done

File: src/databricks/labs/ucx/hive_metastore/table_size.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+    def _crawl(self) -> Iterable[TableSize]:
+        """Crawls and lists tables using table crawler
+        Identifies DBFS root tables and calculates the size for these.
+        """
+        table_sizes = []
+        for table in self._table_crawler.snapshot():
+            if not table.kind == "TABLE" or not table.is_dbfs_root():
+                continue
+            size_in_bytes = self._backend.get_table_size(f"{table.catalog}.{table.database}.{table.name}")
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_size.py:41]: isn't there `full_name` method on table?

File: src/databricks/labs/ucx/hive_metastore/table_size.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+            table_size = TableSize(
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_size.py:42]: yield it?..

File: src/databricks/labs/ucx/hive_metastore/table_size.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        super().__init__(backend, "hive_metastore", schema, "table_size", TableSize)
+        self._table_crawler = TablesCrawler(backend, schema)
+
+    def _crawl(self) -> Iterable[TableSize]:
+        """Crawls and lists tables using table crawler
+        Identifies DBFS root tables and calculates the size for these.
+        """
+        table_sizes = []
+        for table in self._table_crawler.snapshot():
+            if not table.kind == "TABLE" or not table.is_dbfs_root():
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_size.py:39]: I think I got it.

File: tests/unit/hive_metastore/test_tables.py [hunk 15]
```diff
@@ -107,3 +107,12 @@ def test_tables_returning_error_when_describing():
...
     results = tc._crawl()
     assert len(results) == 1
+
+
+def test_is_dbfs_root():
+    assert Table("a", "b", "c", "MANAGED", "DELTA", location="dbfs:/somelocation/tablename").is_dbfs_root()
+    assert Table("a", "b", "c", "MANAGED", "DELTA", location="/dbfs/somelocation/tablename").is_dbfs_root()
+    assert not Table("a", "b", "c", "MANAGED", "DELTA", location="dbfs:/mnt/somelocation/tablename").is_dbfs_root()
+    assert not Table("a", "b", "c", "MANAGED", "DELTA", location="/dbfs/mnt/somelocation/tablename").is_dbfs_root()
+    assert not Table("a", "b", "c", "MANAGED", "DELTA", location="s3:/somelocation/tablename").is_dbfs_root()
```
Comment (nfx) [inline tests/unit/hive_metastore/test_tables.py:117]: add test for databricks datasets

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/framework/crawlers.py (hunk 7 above)
Comment (FastLee) [reply src/databricks/labs/ucx/framework/crawlers.py]: Of course. Fixed.

File: src/databricks/labs/ucx/framework/crawlers.py (hunk 8 above)
Comment (FastLee) [reply src/databricks/labs/ucx/framework/crawlers.py:170]: Table size crawler doesn't have spark context.

File: src/databricks/labs/ucx/hive_metastore/tables.py (hunk 9 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/tables.py]: Do they show up as managed table?

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 16]
```diff
@@ -80,6 +80,23 @@ def sql_unset_upgraded_to(self, catalog):
             f"UNSET TBLPROPERTIES IF EXISTS('upgraded_to');"
         )
 
+    def is_dbfs_root(self) -> bool:
+        if not self.location:
+            return False
+        if (
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/tables.py]: Fixed

File: src/databricks/labs/ucx/framework/crawlers.py (hunk 6 above)
Comment (FastLee) [reply src/databricks/labs/ucx/framework/crawlers.py:172]: Fixed

File: src/databricks/labs/ucx/framework/crawlers.py (hunk 10 above)
Comment (FastLee) [reply src/databricks/labs/ucx/framework/crawlers.py]: done

File: src/databricks/labs/ucx/hive_metastore/table_size.py (hunk 12 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_size.py:42]: nope, but it does have key() function.
Thanks

File: src/databricks/labs/ucx/hive_metastore/table_size.py (hunk 13 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_size.py:43]: Fixed

File: tests/unit/hive_metastore/test_tables.py (hunk 15 above)
Comment (FastLee) [reply tests/unit/hive_metastore/test_tables.py:117]: Databricks datasets are located in  dbfs:/databricks-datasets and will be handled just like any other dbfs root location.

File: src/databricks/labs/ucx/hive_metastore/tables.py (hunk 9 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/tables.py]: The are located in dbfs:/databricks-datasets and will be handled like any other dbfs root location.

File: src/databricks/labs/ucx/framework/crawlers.py (hunk 8 above)
Comment (nfx) [reply src/databricks/labs/ucx/framework/crawlers.py:170]: You get it from spark session you create there.

This method doesn't belong here. There's no way to get it from SQL, so "no" means "no"

File: src/databricks/labs/ucx/hive_metastore/table_size.py (hunk 13 above)
Comment (nfx) [reply src/databricks/labs/ucx/hive_metastore/table_size.py:43]: No, there's no yield of table size yet

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 17]
```diff
@@ -80,6 +80,25 @@ def sql_unset_upgraded_to(self, catalog):
             f"UNSET TBLPROPERTIES IF EXISTS('upgraded_to');"
         )
 
+    def is_dbfs_root(self) -> bool:
+        if self.location is None:
+            return False
+        if self.location.startswith("/dbfs/mnt"):
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/tables.py:86]: Nit: Can you create a "dbfs_root_prefixes" list and iterate over it?

File: src/databricks/labs/ucx/queries/assessment/main/01_3_all_tables.sql [hunk 18]
```diff
@@ -13,5 +13,9 @@ SELECT CONCAT(`database`, '.', name) AS name,
            ELSE "EXTERNAL"
        END AS storage,
        IF(format = "DELTA", "Yes", "No") AS is_delta,
-       location
-FROM $inventory.tables
+       location,
+       size_in_bytes
```
Comment (nfx) [inline src/databricks/labs/ucx/queries/assessment/main/01_3_all_tables.sql:17]: Probably convert it to MB/GB/UNKNOWN with case-when

File: tests/unit/framework/mocks.py (hunk 5 above)
Comment (nfx) [reply tests/unit/framework/mocks.py:58]: Against, you don't need it

File: tests/unit/hive_metastore/test_tables.py (hunk 15 above)
Comment (nfx) [reply tests/unit/hive_metastore/test_tables.py:117]: Just copy the line and add an assertion

**Review by nfx — CHANGES_REQUESTED** [review]

File: tests/unit/hive_metastore/test_tables.py (hunk 15 above)
Comment (FastLee) [reply tests/unit/hive_metastore/test_tables.py:117]: I did.

File: tests/unit/framework/mocks.py (hunk 5 above)
Comment (FastLee) [reply tests/unit/framework/mocks.py:58]: Removed

File: src/databricks/labs/ucx/framework/crawlers.py (hunk 8 above)
Comment (FastLee) [reply src/databricks/labs/ucx/framework/crawlers.py:170]: The spark session is private. Should I make it public?

File: src/databricks/labs/ucx/hive_metastore/table_size.py (hunk 13 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_size.py:43]: Fixed

File: src/databricks/labs/ucx/hive_metastore/tables.py (hunk 17 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/tables.py:86]: Sure, makes sense. I was thinking just that.

File: src/databricks/labs/ucx/framework/crawlers.py (hunk 8 above)
Comment (FastLee) [reply src/databricks/labs/ucx/framework/crawlers.py:170]: Fixed

Comment (nfx) [reply src/databricks/labs/ucx/framework/crawlers.py:170]: Revert this change from here

File: src/databricks/labs/ucx/hive_metastore/table_size.py [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+    def _crawl(self) -> Iterable[TableSize]:
+        """Crawls and lists tables using table crawler
+        Identifies DBFS root tables and calculates the size for these.
+        """
+        for table in self._tables_crawler.snapshot():
+            if not table.kind == "TABLE":
+                continue
+            if not table.is_dbfs_root():
+                continue
+            size_in_bytes = self._backend.get_table_size(table.key)
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/table_size.py:42]: Call spark from here, not from runtime backend.

File: tests/unit/framework/mocks.py (hunk 5 above)
Comment (nfx) [reply tests/unit/framework/mocks.py:58]: This is not removed.

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/queries/assessment/main/01_3_all_tables.sql (hunk 18 above)
Comment (FastLee) [reply src/databricks/labs/ucx/queries/assessment/main/01_3_all_tables.sql]: Done

File: src/databricks/labs/ucx/hive_metastore/table_size.py (hunk 19 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/table_size.py]: Done

File: src/databricks/labs/ucx/queries/assessment/main/01_3_all_tables.sql [hunk 20]
```diff
@@ -1,17 +1,31 @@
...
            WHEN STARTSWITH(location, "adl") THEN "UNSUPPORTED"
            ELSE "EXTERNAL"
        END AS storage,
        IF(format = "DELTA", "Yes", "No") AS is_delta,
-       location
-FROM $inventory.tables
+       location,
+       CASE
+            WHEN size_in_bytes IS null THEN "Non DBFS Root"
+            WHEN size_in_bytes = 9223372036854775807 THEN "INVALID SIZE ESTIMATE"
```
Comment (nfx) [inline src/databricks/labs/ucx/queries/assessment/main/01_3_all_tables.sql:21]:
```suggestion
```
irrelevant

File: src/databricks/labs/ucx/queries/assessment/main/01_3_all_tables.sql [hunk 21]
```diff
@@ -1,17 +1,31 @@
...
+            WHEN size_in_bytes < 100 THEN CONCAT(CAST(size_in_bytes AS string)," Bytes")
```
Comment (nfx) [inline src/databricks/labs/ucx/queries/assessment/main/01_3_all_tables.sql:22]:
```suggestion
            WHEN size_in_bytes < 100 THEN CONCAT(CAST(size_in_bytes AS string),"Bytes")
```
BE CONSISTENT

File: src/databricks/labs/ucx/framework/crawlers.py [hunk 22]
```diff
@@ -147,15 +147,15 @@ def __init__(self):
             msg = "Not in the Databricks Runtime"
             raise RuntimeError(msg)
 
-        self._spark = SparkSession.builder.getOrCreate()
+        self.spark = SparkSession.builder.getOrCreate()
```
Comment (nfx) [inline src/databricks/labs/ucx/framework/crawlers.py:150]: How many times do i have to repeat that you're not allowed to make any change like that to `RuntimeBackend`? Invoke `SparkSession.builder.getOrCreate()` + runtime check from `TableSizes`.

File: tests/unit/hive_metastore/test_tables.py (hunk 15 above)
Comment (nfx) [reply tests/unit/hive_metastore/test_tables.py:117]: where?

**Review by nfx — CHANGES_REQUESTED** [review]
code is incorrect and broken. see comments.

File: tests/unit/framework/mocks.py (hunk 5 above)
Comment (FastLee) [reply tests/unit/framework/mocks.py]: Swear to god it was removed

File: src/databricks/labs/ucx/queries/assessment/main/01_3_all_tables.sql (hunk 20 above)
Comment (FastLee) [reply src/databricks/labs/ucx/queries/assessment/main/01_3_all_tables.sql:21]: Why? There's a known bug.

File: src/databricks/labs/ucx/queries/assessment/main/01_3_all_tables.sql (hunk 21 above)
Comment (FastLee) [reply src/databricks/labs/ucx/queries/assessment/main/01_3_all_tables.sql:22]: It looks funny 88Bytes where 8GB looks right

File: src/databricks/labs/ucx/framework/crawlers.py (hunk 22 above)
Comment (FastLee) [reply src/databricks/labs/ucx/framework/crawlers.py:150]: Fixed

File: src/databricks/labs/ucx/framework/crawlers.py (hunk 8 above)
Comment (FastLee) [reply src/databricks/labs/ucx/framework/crawlers.py]: Fixed

File: tests/unit/hive_metastore/test_tables.py (hunk 15 above)
Comment (FastLee) [reply tests/unit/hive_metastore/test_tables.py:123]: There

File: src/databricks/labs/ucx/queries/assessment/main/01_3_all_tables.sql (hunk 20 above)
Comment (FastLee) [reply src/databricks/labs/ucx/queries/assessment/main/01_3_all_tables.sql]: I set up a cap at 10PB.
Also this error is only with parquet.
We may want to omit Parquet file altogether as we are not "deep cloning" them.
We have to figure how to handle parquet and other formats.

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**